frontend/src/pages/login/oidc-redirect.tsx

```diff
@@ -76,107 +126,289 @@ const OpenIdConnectRedirect = () => {
-        );
+        // this is the first time the teacher is logging in, create a new key pair
 
-        // replace this once the backend is connected
-        const tempTeacherKeyPair = await TeacherLongTermKeyPair.generate(
+        teacherKeyPair = await TeacherLongTermKeyPair.generate(
           window.crypto.subtle,
         );
 
-        const [encryptedPrivateKey, publicKey] = await Promise.all([
-          tempTeacherKeyPair.exportPrivateKey(passwordKey),
-          tempTeacherKeyPair.exportPublicKey(),
-        ]);
+        const salt1 = PasswordDerivedKey.generateSalt();
+        const pwKey1 = await PasswordDerivedKey.derive(
+          crypto,
+          userProvidedPassword,
+          salt1,
+        );
 
-        const teacherKeyPair = await TeacherLongTermKeyPair.importKeyPair(
-          window.crypto.subtle,
-          encryptedPrivateKey,
-          publicKey,
-          passwordKey,
+        const salt2 = PasswordDerivedKey.generateSalt();
+        const pwKey2 = await PasswordDerivedKey.derive(
+          crypto,
+          userProvidedBackupPassword,
+          salt2,
         );
 
-        updateAuthenticationContext({
-          version: latestAuthenticationContextVersion,
-          idToken: idToken,
+        const privateKey1 = await teacherKeyPair.exportPrivateKey(pwKey1);
+        const privateKey2 = await teacherKeyPair.exportPrivateKey(pwKey2);
+
+        const publicKey = await teacherKeyPair.exportPublicKey();
+        const saltPublicKey = await teacherKeyPair.exportSaltPublicKey();
+        const fingerprint = await teacherKeyPair.getPublicKeyFingerprint();
+
+        keyPairId = await updateUserKey(
+          userId,
+          {
+            key: {
+              publicKey: JSON.stringify(publicKey),
+              publicKeyFingerprint: fingerprint,
+              salt: encodeBase64(
+                new TextEncoder().encode(JSON.stringify(saltPublicKey)),
+              ),
+              privateKeys: [
+                {
+                  salt: encodeBase64(salt1),
+                  encryptedPrivateKey: encodeBase64(privateKey1),
+                },
+                {
+                  salt: encodeBase64(salt2),
+                  encryptedPrivateKey: encodeBase64(privateKey2),
+                },
+              ],
+            },
+          },
+          // explicitly set the authentication token because in the local application state
+          // the user is not authenticated yet
           authenticationToken,
-          role,
-          email,
-          name,
-          keyPair: teacherKeyPair,
+        );
+      } else {
+        // the teacher has logged in before, decrypt the private key
+        const { publicKey: publicKeyString, salt, privateKeys } = keyPair;
+        const publicKey = JSON.parse(publicKeyString);
+        const saltPublicKey = JSON.parse(
+          new TextDecoder().decode(decodeBase64(salt)),
+        );
+
+        // try to decrypt all private keys with the password
```

Comment:
this runs the risk that the user forgets the other password :-/

Reply:
Hmm I see what you mean - you only want to try with the first and then - if it doesn't work - try with the second and give the user the option to encrypt a new key etc. That makes sense - https://classroomrt.atlassian.net/browse/CRT-55.